packages/peregrine/lib/talons/RootComponents/Product/useProduct.js

```diff
@@ -0,0 +1,111 @@
+import { useEffect, useMemo } from 'react';
+import { useApolloClient, useLazyQuery } from '@apollo/react-hooks';
+
+/**
+ * A [React Hook]{@link https://reactjs.org/docs/hooks-intro.html} that
+ * controls the logic for the Product Root Component.
+ *
+ * @kind function
+ *
+ * @param {object} props
+ *
+ */
+export const useProduct = props => {
+    const {
+        fragment,
+        query,
+        urlKey // TODO: move `getUrlKey` into Peregrine?
+    } = props;
+
+    const onServer = false;
+
+    const fragmentVariables = useMemo(() => {
+        return { onServer };
+    }, [onServer]);
+
+    const queryVariables = useMemo(() => {
+        return {
+            onServer,
+            urlKey
+        };
+    }, [onServer, urlKey]);
+
+    const [runQuery, { loading, error, data }] = useLazyQuery(query, {
+        // We're manually checking the cache here, so useLazyQuery can skip that.
+        fetchPolicy: 'network-only',
+        variables: queryVariables
+    });
+
+    const apolloClient = useApolloClient();
+
+    const productFromCache = useMemo(() => {
+        /*
+         * Look up the product in cache first.
+         *
+         * We may have it from a previous query, but we have to manually tell Apollo cache where to find it.
+         * A single product query (as described by https://github.com/magento/graphql-ce/issues/86)
+         * will alleviate the need to do this manual work.
+         *
+         * If the object with the specified `id` is not in the cache, we get `null`.
+         * If the object is in the cache but doesn't have all the fields we need, an error is thrown.
+         *
+         * @see https://www.apollographql.com/docs/react/caching/cache-interaction/#readfragment.
+         */
+        try {
+            return apolloClient.readFragment({
+                // This `id` must match the result of `cacheKeyFromType` in `venia-ui/lib/util/apolloCache.js`.
+                id: `ProductInterface:${urlKey}`,
+                fragment,
+                variables: fragmentVariables
+            });
+        } catch (e) {
+            // The product is in the cache but it is missing some fields the fragment needs.
+            // We don't have to do anything here.
+            return null;
+        }
+    }, [apolloClient, fragment, fragmentVariables, urlKey]);
+
+    const product = useMemo(() => {
+        const mapProduct = product => {
```

Comment:
This can/should exist out of the product talon so we aren't recreating it (even though it's semi protected by memoization).

Reply:
Resolved by https://github.com/magento/pwa-studio/pull/1935/commits/28de28a26c2ed22153a2c25666b73bcdedc5b9f7.